### PR TITLE
[ALLUXIO-2816] Use domain socket to do short circuit read  

### PR DESCRIPTION
**Pull request by peisun1115**

### CODE REVIEW

**Comment by peisun1115** [comment]
@calvinjia @gpang This PR is ready for review. I did not remove short circuit read in this PR.

File: tests/src/test/java/alluxio/master/journal/ufs/UfsJournalIntegrationTest.java [hunk 1]
```diff
@@ -86,6 +87,7 @@ public final void before() throws Exception {
    * Tests adding a block.
    */
   @Test
+  @Ignore
```
Comment (gpang) [inline tests/src/test/java/alluxio/master/journal/ufs/UfsJournalIntegrationTest.java]: why are all of these ignored?

File: integration/docker/Dockerfile [hunk 2]
```diff
@@ -11,19 +11,22 @@
 
 FROM openjdk:7-jdk-alpine
 
-ARG ALLUXIO_TARBALL=http://downloads.alluxio.org/downloads/files/1.4.0/alluxio-1.4.0-bin.tar.gz
+# ARG ALLUXIO_TARBALL=http://downloads.alluxio.org/downloads/files/1.4.0/alluxio-1.4.0-bin.tar.gz
```
Comment (gpang) [inline integration/docker/Dockerfile]: why are these changes required?

File: tests/src/test/java/alluxio/master/journal/ufs/UfsJournalIntegrationTest.java (hunk 1 above)
Comment (calvinjia) [inline tests/src/test/java/alluxio/master/journal/ufs/UfsJournalIntegrationTest.java]: Revert changes in this file?

File: core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannelV2.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+  @Override
+  public long getLength() {
+    return mLength;
+  }
+
+  @Override
+  public ByteBuffer getReadOnlyByteBuffer() {
+    throw new UnsupportedOperationException(
+        "DataFileChannel#getReadOnlyByteBuffer is not implemented.");
```
Comment (gpang) [inline core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannelV2.java]: "DataFileChannelV2"

File: core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannelV2.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  @Override
+  public void readBytes(byte[] dst, int dstIndex, int length) {
+    throw new UnsupportedOperationException("DataFileChannel#readBytes is not implemented.");
```
Comment (gpang) [inline core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannelV2.java]: "DataFileChannelV2"

File: core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannelV2.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  @Override
+  public int readableBytes() {
+    throw new UnsupportedOperationException("DataFileChannel#readableBytes is not implemented.");
```
Comment (gpang) [inline core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannelV2.java]: "DataFileChannelV2"

File: core/common/src/main/java/alluxio/PropertyKey.java [hunk 6]
```diff
@@ -448,7 +449,7 @@
         "alluxio.zookeeper.leader.inquiry.retry";
     public static final String ZOOKEEPER_LEADER_PATH = "alluxio.zookeeper.leader.path";
 
-    //
+   //
```
Comment (gpang) [inline core/common/src/main/java/alluxio/PropertyKey.java]: wrong indent?

File: core/common/src/main/java/alluxio/PropertyKey.java [hunk 7]
```diff
@@ -607,6 +608,8 @@
     public static final String WORKER_DATA_HOSTNAME = "alluxio.worker.data.hostname";
     public static final String WORKER_DATA_PORT = "alluxio.worker.data.port";
     public static final String WORKER_DATA_SERVER_CLASS = "alluxio.worker.data.server.class";
+    public static final String WORKER_DATA_SERVER_DOMAIN_SOCKET_ADDRESS =
+        "alluxio.worker.data.server.domain.socket.address";
```
Comment (gpang) [inline core/common/src/main/java/alluxio/PropertyKey.java]: Does a worker have only a single domain socket? Are there any implications of having only 1 domain socket vs more than 1?

File: core/client/src/main/java/alluxio/client/netty/NettyClient.java [hunk 8]
```diff
@@ -56,31 +59,29 @@
       Configuration.getInt(PropertyKey.USER_NETWORK_NETTY_WORKER_THREADS), "netty-client-worker-%d",
       true);
 
-  /** The maximum number of milliseconds to wait for a response from the server. */
-  public static final long TIMEOUT_MS =
```
Comment (gpang) [inline core/client/src/main/java/alluxio/client/netty/NettyClient.java]: where is the timeout now?

File: core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java [hunk 9]
```diff
@@ -227,11 +227,11 @@ public void start(WorkerNetAddress address) {
    */
   @Override
   public void stop() {
+    // Use shutdownNow because HeartbeatThreads never finish until they are interrupted
+    getExecutorService().shutdownNow();
```
Comment (gpang) [inline core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java]: Is this a repeat of the fix from the FT test PR?

File: core/client/src/main/java/alluxio/client/netty/NettyClient.java (hunk 8 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/netty/NettyClient.java]: This was not used after non-packet streaming code is removed. It is now defined in NettyPacketReader/Writer.

File: core/common/src/main/java/alluxio/PropertyKey.java (hunk 7 above)
Comment (peisun1115) [inline core/common/src/main/java/alluxio/PropertyKey.java]: I do not see any reason to have more than 1. AFAIK, this address plays similar role as port.

File: core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannelV2.java (hunk 3 above)
Comment (peisun1115) [inline core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannelV2.java]: done

File: core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannelV2.java (hunk 4 above)
Comment (peisun1115) [inline core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannelV2.java]: done

File: core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannelV2.java (hunk 5 above)
Comment (peisun1115) [inline core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannelV2.java]: done

File: core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java (hunk 9 above)
Comment (peisun1115) [inline core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java]: Done. My git branches were messed up

File: integration/docker/Dockerfile (hunk 2 above)
Comment (peisun1115) [inline integration/docker/Dockerfile]: removed. used for testing.

File: tests/src/test/java/alluxio/master/journal/ufs/UfsJournalIntegrationTest.java (hunk 1 above)
Comment (peisun1115) [inline tests/src/test/java/alluxio/master/journal/ufs/UfsJournalIntegrationTest.java]: reverted.

Comment (peisun1115) [inline tests/src/test/java/alluxio/master/journal/ufs/UfsJournalIntegrationTest.java]: done

File: core/server/worker/src/main/java/alluxio/worker/AlluxioWorkerProcess.java [hunk 10]
```diff
@@ -212,6 +233,9 @@ private void stopWorkers() throws Exception {
 
   private void stopServing() throws IOException {
     mDataServer.close();
+    if (mDomainSocketDataServer != null) {
+      mDomainSocketDataServer.close();
```
Comment (gpang) [inline core/server/worker/src/main/java/alluxio/worker/AlluxioWorkerProcess.java]: set `mDomainSocketDataServer ` to `null`?

File: core/common/src/main/java/alluxio/PropertyKey.java (hunk 7 above)
Comment (gpang) [inline core/common/src/main/java/alluxio/PropertyKey.java]: ACK

File: core/client/src/main/java/alluxio/client/netty/NettyClient.java (hunk 8 above)
Comment (gpang) [inline core/client/src/main/java/alluxio/client/netty/NettyClient.java]: ACK

File: core/common/src/main/java/alluxio/network/protocol/RPCProtoMessage.java [hunk 11]
```diff
@@ -60,7 +60,7 @@
   public RPCProtoMessage(ProtoMessage message, DataBuffer data) {
     if (data != null) {
       Preconditions
-          .checkArgument((data instanceof DataNettyBufferV2) || (data instanceof DataFileChannel),
+          .checkArgument((data instanceof DataNettyBufferV2) || (data instanceof DataFileChannelV2),
```
Comment (gpang) [inline core/common/src/main/java/alluxio/network/protocol/RPCProtoMessage.java]: Will there be any backwards compatibility issues?

File: core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannelV2.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+import io.netty.channel.DefaultFileRegion;
+
+import java.io.File;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+
+/**
+ * A DataBuffer with the underlying data being a {@link FileChannel}.
+ */
+public final class DataFileChannelV2 implements DataBuffer {
```
Comment (gpang) [inline core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannelV2.java]: How does this class work? All the reading methods are unsupported.

File: core/common/src/main/java/alluxio/util/network/NettyUtils.java [hunk 13]
```diff
@@ -66,9 +69,16 @@ public static EventLoopGroup createEventLoop(ChannelType type, int numThreads,
...
-   * @return ServerSocketChannel matching the ChannelType
+   * @param isDomainSocket whether this is a domain socket server
+   * @return ServerSocketChannel matching the requirements
    */
-  public static Class<? extends ServerChannel> getServerChannelClass(ChannelType type) {
+  public static Class<? extends ServerChannel> getServerChannelClass(ChannelType type,
+      boolean isDomainSocket) {
+    if (isDomainSocket) {
+      Preconditions.checkState(type == ChannelType.EPOLL,
+          "Domain socket is used while the channel type is not epoll.");
```
Comment (gpang) [inline core/common/src/main/java/alluxio/util/network/NettyUtils.java]: Would "Domain sockets are only supported with EPOLL channel type" be more clear?

File: core/common/src/main/java/alluxio/util/network/NettyUtils.java [hunk 14]
```diff
@@ -83,9 +93,16 @@ public static EventLoopGroup createEventLoop(ChannelType type, int numThreads,
...
-   * @return SocketChannel matching the ChannelType
+   * @param isDomainSocket whether this is to connect to a domain socket server
+   * @return Channel matching the requirements
    */
-  public static Class<? extends SocketChannel> getClientChannelClass(ChannelType type) {
+  public static Class<? extends Channel> getClientChannelClass(ChannelType type,
+      boolean isDomainSocket) {
+    if (isDomainSocket) {
+      Preconditions.checkState(type == ChannelType.EPOLL,
+          "Domain socket is used while the channel type is not epoll.");
```
Comment (gpang) [inline core/common/src/main/java/alluxio/util/network/NettyUtils.java]: Would "Domain sockets are only supported with EPOLL channel type" be more clear?

File: core/server/worker/src/main/java/alluxio/worker/netty/NettyDataServer.java [hunk 15]
```diff
@@ -170,8 +163,8 @@ private ServerBootstrap createBootstrapOfType(final ChannelType type) {
     final EventLoopGroup workerGroup =
         NettyUtils.createEventLoop(type, workerThreadCount, "data-server-worker-%d", false);
```
Comment (gpang) [inline core/server/worker/src/main/java/alluxio/worker/netty/NettyDataServer.java]: Do both the network server and the domain socket server both use this "data-server-worker-%d" name?

File: core/common/src/main/java/alluxio/PropertyKey.java [hunk 16]
```diff
@@ -181,6 +181,7 @@
   WORKER_DATA_HOSTNAME(Name.WORKER_DATA_HOSTNAME, null),
   WORKER_DATA_PORT(Name.WORKER_DATA_PORT, 29999),
   WORKER_DATA_SERVER_CLASS(Name.WORKER_DATA_SERVER_CLASS, "alluxio.worker.netty.NettyDataServer"),
+  WORKER_DATA_SERVER_DOMAIN_SOCKET_ADDRESS(Name.WORKER_DATA_SERVER_DOMAIN_SOCKET_ADDRESS, ""),
```
Comment (gpang) [inline core/common/src/main/java/alluxio/PropertyKey.java]: Update documentation.

File: core/common/src/main/java/alluxio/network/protocol/RPCProtoMessage.java (hunk 11 above)
Comment (peisun1115) [inline core/common/src/main/java/alluxio/network/protocol/RPCProtoMessage.java]: No.  This change does not affect the actual data transferred

File: core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannelV2.java (hunk 12 above)
Comment (peisun1115) [inline core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannelV2.java]: RPCMessageEncoder#encode calls getNettyOutput which transfers data from the file region to network.

File: core/server/worker/src/main/java/alluxio/worker/netty/NettyDataServer.java (hunk 15 above)
Comment (peisun1115) [inline core/server/worker/src/main/java/alluxio/worker/netty/NettyDataServer.java]: done

File: core/common/src/main/java/alluxio/util/network/NettyUtils.java (hunk 14 above)
Comment (peisun1115) [inline core/common/src/main/java/alluxio/util/network/NettyUtils.java]: done

File: core/common/src/main/java/alluxio/util/network/NettyUtils.java (hunk 13 above)
Comment (peisun1115) [inline core/common/src/main/java/alluxio/util/network/NettyUtils.java]: done

File: core/common/src/main/java/alluxio/PropertyKey.java (hunk 16 above)
Comment (peisun1115) [inline core/common/src/main/java/alluxio/PropertyKey.java]: done

File: core/server/worker/src/main/java/alluxio/worker/AlluxioWorkerProcess.java (hunk 10 above)
Comment (peisun1115) [inline core/server/worker/src/main/java/alluxio/worker/AlluxioWorkerProcess.java]: done

**Comment by peisun1115** [comment]
@gpang Thanks for the review. PTAL.

**Comment by peisun1115** [comment]
@gpang @calvinjia Do you have any comments?

File: core/client/src/main/java/alluxio/client/block/StreamFactory.java [hunk 17]
```diff
@@ -43,6 +44,10 @@ private StreamFactory() {} // prevent instantiation
    */
   public static BlockOutStream createLocalBlockOutStream(FileSystemContext context, long blockId,
       long blockSize, WorkerNetAddress address, OutStreamOptions options) {
+    if (NettyClient.isDomainSocketSupported(address)) {
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/block/StreamFactory.java]: I would prefer to keep the Configuration based methods outside of `NettyClient` and have it be a factory for creating client bootstraps.

File: core/client/src/main/java/alluxio/client/block/StreamFactory.java [hunk 18]
```diff
@@ -43,6 +44,10 @@ private StreamFactory() {} // prevent instantiation
...
+      return BlockOutStream
+          .createNettyBlockOutStream(blockId, blockSize, address, context, options);
+    }
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/block/StreamFactory.java]: The descriptor `create*BlockOutStream` is no longer symmetrical

File: core/client/src/main/java/alluxio/client/block/stream/BlockInStream.java [hunk 19]
```diff
@@ -101,19 +104,24 @@ public static BlockInStream createLocalBlockInStream(long blockId, long blockSiz
...
+  public static BlockInStream createNettyBlockInStream(long blockId, long blockSize,
       WorkerNetAddress workerNetAddress, FileSystemContext context, InStreamOptions options) {
     Closer closer = Closer.create();
     try {
       BlockWorkerClient blockWorkerClient =
           closer.register(context.createBlockWorkerClient(workerNetAddress));
       LockBlockResource lockBlockResource =
           closer.register(blockWorkerClient.lockBlock(blockId, LockBlockOptions.defaults()));
+      SocketAddress address;
+      if (NettyClient.isDomainSocketSupported(workerNetAddress)) {
```
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/block/stream/BlockInStream.java]: This is called many times in the code path?

File: core/server/worker/src/main/java/alluxio/worker/block/BlockMasterClient.java [hunk 20]
```diff
@@ -96,8 +96,9 @@ public synchronized long getId(final WorkerNetAddress address) {
     return retryRPC(new RpcCallable<Long>() {
       @Override
       public Long call() throws TException {
-        return mClient.getWorkerId(new alluxio.thrift.WorkerNetAddress(address.getHost(),
-            address.getRpcPort(), address.getDataPort(), address.getWebPort()));
+        return mClient.getWorkerId(
+            new alluxio.thrift.WorkerNetAddress(address.getHost(), address.getRpcPort(),
+                address.getDataPort(), address.getWebPort(), address.getDomainSocketPath()));
```
Comment (calvinjia) [inline core/server/worker/src/main/java/alluxio/worker/block/BlockMasterClient.java]: Can we use `toThrift` here?

File: core/common/src/main/java/alluxio/util/network/NettyUtils.java [hunk 21]
```diff
@@ -66,9 +69,16 @@ public static EventLoopGroup createEventLoop(ChannelType type, int numThreads,
    * {@link ChannelType}.
    *
    * @param type Selector for which form of low-level IO we should use
-   * @return ServerSocketChannel matching the ChannelType
+   * @param isDomainSocket whether this is a domain socket server
+   * @return ServerSocketChannel matching the requirements
    */
-  public static Class<? extends ServerChannel> getServerChannelClass(ChannelType type) {
+  public static Class<? extends ServerChannel> getServerChannelClass(ChannelType type,
+      boolean isDomainSocket) {
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/util/network/NettyUtils.java]: Instead of having the boolean, can we use another `ChannelType`? (applies to below as well)

File: core/server/worker/src/main/java/alluxio/worker/netty/NettyDataServer.java [hunk 22]
```diff
@@ -165,13 +158,17 @@ private ServerBootstrap createBootstrapOfType(final ChannelType type) {
...
-    final EventLoopGroup bossGroup =
-        NettyUtils.createEventLoop(type, bossThreadCount, "data-server-boss-%d", false);
-    final EventLoopGroup workerGroup =
-        NettyUtils.createEventLoop(type, workerThreadCount, "data-server-worker-%d", false);
-
-    final Class<? extends ServerChannel> socketChannelClass =
-        NettyUtils.getServerChannelClass(type);
+    String dataServerEventLoopNamePrefix =
+        "data-server-" + ((mSocketAddress instanceof DomainSocketAddress) ? "domain-socket" :
+            "tcp-socket");
```
Comment (calvinjia) [inline core/server/worker/src/main/java/alluxio/worker/netty/NettyDataServer.java]: We either expect threads of one kind or the other but never both, so is it important to distinguish?

File: core/common/src/main/java/alluxio/network/protocol/RPCProtoMessage.java (hunk 11 above)
Comment (calvinjia) [inline core/common/src/main/java/alluxio/network/protocol/RPCProtoMessage.java]: Could you update the error message below?

File: core/common/src/main/java/alluxio/util/network/NettyUtils.java (hunk 21 above)
Comment (peisun1115) [inline core/common/src/main/java/alluxio/util/network/NettyUtils.java]: I have thought about it. The problem is that EPOLL and DOMAIN_SOCKET are not mutual exclusive. 

EPOLL and NIO are different ways to poll on sockets. but Domain socket is a different socket. It also sounds strange to put domain socket as something comparable with EPOLL and NIO.

File: core/server/worker/src/main/java/alluxio/worker/block/BlockMasterClient.java (hunk 20 above)
Comment (peisun1115) [inline core/server/worker/src/main/java/alluxio/worker/block/BlockMasterClient.java]: done.

File: core/client/src/main/java/alluxio/client/block/StreamFactory.java (hunk 17 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/block/StreamFactory.java]: I moved this to NettyUtils.

File: core/client/src/main/java/alluxio/client/block/StreamFactory.java (hunk 18 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/block/StreamFactory.java]: yes. any suggestion?

File: core/client/src/main/java/alluxio/client/block/stream/BlockInStream.java (hunk 19 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/block/stream/BlockInStream.java]: I can optionally pass the result this call from StreamFactory. But i thought that this call is cheap enough (a string comparison) and didn't do that. Let me know if you prefer more params to be passed.

File: core/client/src/main/java/alluxio/client/block/StreamFactory.java (hunk 18 above)
Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/block/StreamFactory.java]: I think `Local` no longer makes sense since the Netty stream with domain socket could also be reading from a local machine.

File: core/server/worker/src/main/java/alluxio/worker/netty/NettyDataServer.java (hunk 22 above)
Comment (calvinjia) [inline core/server/worker/src/main/java/alluxio/worker/netty/NettyDataServer.java]: you're right, we need both, disregard this comment.

File: core/common/src/main/java/alluxio/util/network/NettyUtils.java (hunk 21 above)
Comment (calvinjia) [inline core/common/src/main/java/alluxio/util/network/NettyUtils.java]: Would `EPOLL_DOMAIN_SOCKET` work? The logic in the method seems to just be mapping ChannelType to a ServerChannel class.

File: core/client/src/main/java/alluxio/client/block/StreamFactory.java (hunk 18 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/block/StreamFactory.java]: I've renamed "local" to "short circuit" in Block{In,Out}Stream but left the names in StreamFactory unchanged. Let me know that is not what you want.

Comment (calvinjia) [inline core/client/src/main/java/alluxio/client/block/StreamFactory.java]: LGTM, I think the local short circuit can read from non ramdisk files as well?

File: core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannelV2.java (hunk 12 above)
Comment (calvinjia) [inline core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannelV2.java]: Is V2 a replacement for `DataFileChannel`?

Comment (peisun1115) [inline core/common/src/main/java/alluxio/network/protocol/databuffer/DataFileChannelV2.java]: Yes. I've removed DataFileChannel which is no longer necessary since non-packet streaming code is removed.

File: core/common/src/main/java/alluxio/util/network/NettyUtils.java (hunk 21 above)
Comment (peisun1115) [inline core/common/src/main/java/alluxio/util/network/NettyUtils.java]: ack.

**Comment by peisun1115** [comment]
@calvinjia @gpang  PTAL?
